src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs

```diff
@@ -141,80 +142,29 @@ public sealed class ProjectInfo
             ProjectAttributes attributes,
             CompilationOptions? compilationOptions,
             ParseOptions? parseOptions,
-            IEnumerable<DocumentInfo>? documents,
-            IEnumerable<ProjectReference>? projectReferences,
-            IEnumerable<MetadataReference>? metadataReferences,
-            IEnumerable<AnalyzerReference>? analyzerReferences,
-            IEnumerable<DocumentInfo>? additionalDocuments,
-            IEnumerable<DocumentInfo>? analyzerConfigDocuments,
+            IReadOnlyList<DocumentInfo> documents,
+            IReadOnlyList<ProjectReference> projectReferences,
+            IReadOnlyList<MetadataReference> metadataReferences,
+            IReadOnlyList<AnalyzerReference> analyzerReferences,
+            IReadOnlyList<DocumentInfo> additionalDocuments,
+            IReadOnlyList<DocumentInfo> analyzerConfigDocuments,
             Type? hostObjectType)
         {
             Attributes = attributes;
             CompilationOptions = compilationOptions;
             ParseOptions = parseOptions;
-            Documents = documents.ToImmutableReadOnlyListOrEmpty();
-            ProjectReferences = projectReferences.ToImmutableReadOnlyListOrEmpty();
-            MetadataReferences = metadataReferences.ToImmutableReadOnlyListOrEmpty();
-            AnalyzerReferences = analyzerReferences.ToImmutableReadOnlyListOrEmpty();
-            AdditionalDocuments = additionalDocuments.ToImmutableReadOnlyListOrEmpty();
-            AnalyzerConfigDocuments = analyzerConfigDocuments.ToImmutableReadOnlyListOrEmpty();
+            Documents = documents;
+            ProjectReferences = projectReferences;
+            MetadataReferences = metadataReferences;
+            AnalyzerReferences = analyzerReferences;
+            AdditionalDocuments = additionalDocuments;
+            AnalyzerConfigDocuments = analyzerConfigDocuments;
```

Comment:
this feels broken.  Won't this allow someone to pass something *mutbale* that they are exposing through an IReadOnlyList wrapper?  I would expect we would have to only allow that if it was an actual IMmutableArray, and we would still need protection/copying to be resilient to misbehaved callers.

Reply:
we should actually test for this case as well.

Reply:
This is a private ctor called from `Create` method that does the conversion to immutable array.

Reply:
Added a test.